Release/include/cpprest/http_compression.h

```diff
@@ -0,0 +1,329 @@
+/// The supplied string is copied, and thus need not remain valid once the call returns.</param>
+/// <param name="make_compressor">A factory function to be used to instantiate a compressor matching the factory's
+/// reported algorithm.</param>
+/// <returns>
+/// A pointer to a generic provider factory implementation configured with the supplied parameters.
+/// </returns>
+/// <remarks>
+/// This method may be used to conveniently instantiate a factory object for a caller-selected <c>compress_provider</c>.
+/// That provider may be of the caller's own design, or it may be one of the built-in types.  As such, this method may
+/// be helpful when a caller wishes to build vectors containing a mix of custom and built-in providers.
+/// </remarks>
+_ASYNCRTIMP std::shared_ptr<compress_factory> make_compress_factory(
+    const utility::string_t& algorithm, std::function<std::unique_ptr<compress_provider>()> make_compressor);
+
+/// <summary>
+/// Factory function to instantiate a decompression provider factory by compression algorithm name.
+/// </summary>
+/// <param name="algorithm">The name of the algorithm supported by the factory.  Must match that returned by the
+/// <c>web::http::compression::decompress_provider</c> type instantiated by the factory's make_decompressor function.
+/// The supplied string is copied, and thus need not remain valid once the call returns.</param>
+/// <param name="weight">A numeric weight for the compression algorithm, times 1000,  for use as a "quality value" when
+/// requesting that the server send a compressed response.  Valid values are between 0 and 1000, inclusive, where higher
+/// values indicate more preferred algorithms, and 0 indicates that the algorithm is not allowed; values greater than
+/// 1000 are treated as 1000.</param>
+/// <param name="make_decompressor">A factory function to be used to instantiate a decompressor matching the factory's
+/// reported algorithm.</param>
+/// <returns>
+/// A pointer to a generic provider factory implementation configured with the supplied parameters.
+/// </returns>
+/// <remarks>
+/// This method may be used to conveniently instantiate a factory object for a caller-selected
+/// <c>decompress_provider</c>.  That provider may be of the caller's own design, or it may be one of the built-in
+/// types.  As such, this method may be helpful when a caller wishes to change the weights of built-in provider types,
+/// to use custom providers without explicitly implementing a <c>decompress_factory</c>, or to build vectors containing
+/// a mix of custom and built-in providers.
+/// </remarks>
+_ASYNCRTIMP std::shared_ptr<decompress_factory> make_decompress_factory(
+    const utility::string_t& algorithm,
+    uint16_t weight,
+    std::function<std::unique_ptr<decompress_provider>()> make_decompressor);
+
+namespace details
+{
+namespace builtin
+{
+// Internal-only helper function
```

Comment:
It's in the header so that `http_request::set_decompress_factories()` in http_msg.cpp can use it to set its `m_compressor` member to the full set of builtins.  It's "internal" in the sense of "internal to the cpprestsdk library" rather than "internal to the compression module".

It looks like a better approach (with precedent!) would be to move the declaration to internal_http_helpers.h so that it's not exposed in the published headers.  I could add an internal_http_**compression**.h if you think's worth it, but that seems like overkill to me for a few lines -- and internal_http_helpers.h is **already** included by http_msg.cpp.

Reply:
The concern is that we don't want it being a "published" API that people can depend on if it's an implementation detail.

The couple of places we have no choice but to put things in headers we usually use a details namespace.

Reply:
This one **is** in a details namespace, though it's `details::builtin` to denote that it is related to those in the regular `builtin` namespace.

I think it's cleaner to move this one to the internal header, though, as it's otherwise the only API in this one that isn't _ASYNCRTIMP.